File: web/src/lib/notes/localSaveResult.ts
/**
 * Outcome of writing the editor's contents to IndexedDB.
 *
 * Local durability and cloud sync are separate concerns, and conflating them is how a network
 * failure ends up reported as a lost note. Only IndexedDB accepting the write produces `saved`;
 * an attachment upload or Supabase push failing afterwards leaves the save `saved` and is
 * surfaced as pending sync instead.
 */
export type LocalSaveResult =
  | { status: 'saved'; noteId: string }
  | { status: 'unchanged' }
  | { status: 'failed'; error: unknown };

File: composeApp/src/commonMain/kotlin/com/aus/notelikeus/ui/editor/LocalSaveResult.kt
package com.aus.notelikeus.ui.editor

/**
 * Outcome of writing the editor's contents to the local database.
 *
 * Local durability and cloud synchronisation are separate concerns, and conflating them is how a
 * network failure ends up reported as a lost note. Only Room deciding the write landed produces
 * [Saved]; an attachment upload or note sync failing afterwards leaves the save [Saved] and is
 * surfaced as pending sync instead.
 */
sealed interface LocalSaveResult {
    /** Room committed the note. [noteId] is stable and safe to keep in editor state. */
    data class Saved(val noteId: Long) : LocalSaveResult

    /** Nothing needed writing — the editor holds no content. */
    data object Unchanged : LocalSaveResult

    /** The local write failed. The editor still holds the only copy of the edit. */
    data class Failed(val cause: Throwable) : LocalSaveResult
}
